File: svc-position/src/main/java/com/beancounter/position/utils/FxUtils.java
package com.beancounter.position.utils;

import com.beancounter.common.contracts.FxRequest;
import com.beancounter.common.model.Currency;
import com.beancounter.common.model.IsoCurrencyPair;
import com.beancounter.common.model.Position;
import com.beancounter.common.model.Positions;
import com.beancounter.common.utils.DateUtils;
import org.springframework.stereotype.Service;

@Service
public class FxUtils {
  private DateUtils dateUtils = new DateUtils();

  public FxRequest buildRequest(Currency base, Positions positions) {
    if (positions.getAsAt() == null) {
      positions.setAsAt(dateUtils.today());
    }

    FxRequest fxRequest = FxRequest.builder()
        .rateDate(positions.getAsAt())
        .build();

    Currency portfolio = positions.getPortfolio().getCurrency();
    for (Position position : positions.getPositions().values()) {
      fxRequest.add(IsoCurrencyPair.from(base, position.getAsset().getMarket().getCurrency()));
      fxRequest.add(IsoCurrencyPair.from(portfolio, position.getAsset().getMarket().getCurrency()));
    }
    return fxRequest;
  }

}

File: svc-position/src/main/kotlin/com/beancounter/position/utils/FxUtils.kt
package com.beancounter.position.utils

import com.beancounter.common.contracts.FxRequest
import com.beancounter.common.model.Currency
import com.beancounter.common.model.IsoCurrencyPair.Companion.toPair
import com.beancounter.common.model.Position
import com.beancounter.common.model.Positions
import org.springframework.stereotype.Service

/**
 * Helper to build an FxRequest from the supplied arguments.
 */
@Service
class FxUtils {
    fun buildRequest(
        base: Currency,
        positions: Positions,
    ): FxRequest {
        val fxRequest = FxRequest(positions.asAt)
        val portfolio = positions.portfolio.currency
        for (position in positions.positions.values) {
            fxRequest.add(
                toPair(
                    position.getMoneyValues(Position.In.TRADE).currency,
                    base,
                ),
            )
            fxRequest.add(
                toPair(
                    position.getMoneyValues(Position.In.TRADE).currency,
                    portfolio,
                ),
            )
        }
        return fxRequest
    }
}
